servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/ConnectionPoolPolicy.java

```diff
@@ -15,136 +15,22 @@
-     * @param corePoolSize the size of the core pool.
-     * @param forceCorePool whether to avoid selecting connections from the core pool until it has reached the
-     *                      configured core pool size.
-     * @return the configured {@link ConnectionPoolPolicy}.
-     */
-    public static ConnectionPoolPolicy p2c(int corePoolSize, boolean forceCorePool) {
-        return p2c(DEFAULT_MAX_EFFORT, corePoolSize, forceCorePool);
-    }
-
-    /**
-     * A {@link ConnectionPoolPolicy} that attempts to discern between the health of individual connections.
-     * If individual connections have health data the P2C policy can be used to bias traffic toward the best
-     * connections. This has the following algorithm:
-     * - Randomly select two connections from the 'core pool' (pick-two).
-     *   - Try to select the 'best' of the two connections.
-     *   - If we fail to select the best connection, try the other connection.
-     * - If both connections fail, repeat the pick-two operation for up to maxEffort attempts, begin linear iteration
-     *   through the remaining connections searching for an acceptable connection.
-     * @param maxEffort the maximum number of attempts to pick a healthy connection from the core pool.
-     * @param corePoolSize the size of the core pool.
-     * @param forceCorePool whether to avoid selecting connections from the core pool until it has reached the
-     *                      configured core pool size.
-     * @return the configured {@link ConnectionPoolPolicy}.
-     */
-    public static ConnectionPoolPolicy p2c(int maxEffort, int corePoolSize, boolean forceCorePool) {
-        return new P2CPolicy(maxEffort, corePoolSize, forceCorePool);
-    }
-
-    // instance types
-    static final class CorePoolPolicy extends ConnectionPoolPolicy {
-        final int corePoolSize;
-        final boolean forceCorePool;
-
-        CorePoolPolicy(final int corePoolSize, final boolean forceCorePool) {
-            this.corePoolSize = ensurePositive(corePoolSize, "corePoolSize");
-            this.forceCorePool = forceCorePool;
-        }
-    }
-
-    static final class P2CPolicy extends ConnectionPoolPolicy {
-        final int maxEffort;
-        final int corePoolSize;
-        final boolean forceCorePool;
-
-        P2CPolicy(final int maxEffort, final int corePoolSize, final boolean forceCorePool) {
-            this.maxEffort = ensurePositive(maxEffort, "maxEffort");
-            this.corePoolSize = ensurePositive(corePoolSize, "corePoolSize");
-            this.forceCorePool = forceCorePool;
-        }
-    }
-
-    static final class LinearSearchPolicy extends ConnectionPoolPolicy {
-        final int linearSearchSpace;
-
-        LinearSearchPolicy(int linearSearchSpace) {
-            this.linearSearchSpace = ensurePositive(linearSearchSpace, "linearSearchSpace");
-        }
-    }
+    abstract ConnectionSelector<C> buildConnectionSelector(String lbDescription);
```

Comment:
Do you think we are ready to open this up together with `ConnectionSelector`? If yes, we will be able to change from an abstract class to an interface (could be a follow up, no need to mix all in this PR).

Reply:
Good question, and I thought about it as well. The same question applies to the `LoadBalancingPolicy`.

I'm torn: it does feel like it should be an interface but otoh this pattern lets us not commit to any specific API. Let's discuss it in a followup.